code/modules/mining/gemstones.dm

```diff
@@ -0,0 +1,639 @@
+	icon = 'icons/obj/gemstones.dmi'
+	var/gemcolor
+	var/gemtype
+	var/gemsize
+
+/obj/item/precious/gemstone/Initialize(mapload, amount)
+	. = ..()
+	if(amount)
+		stacksize = amount
+		update_gem()
+
+/obj/item/precious/gemstone/mixed/Initialize(mapload, amount, kind)
+	ASSERT(kind)
+	stack_contents[kind] = amount
+	. = ..(mapload, amount)
+
+/obj/item/precious/gemstone/attackby(obj/item/precious/gemstone/G, mob/user)
+	if (!istype(G))
+		return ..()
+	if(G.gemtype == GEM_MIXED)
+		to_chat(user, span("notice", "Try doing it the other way - put an unmixed gem into the pile, instead of the pile onto an organized gem stack."))
+		return
+	if(src.gemsize != G.gemsize)
+		to_chat(user, span("notice", "You can't mix gem sizes. It would be too disorganized!"))
+		return
+	if(src.gemsize == GEM_ENORMOUS)
+		to_chat(user, span("notice", "That gem is far too big to stack!"))
+		return
+	if(src.gemtype == G.gemtype)
+		gemstacker(G, user)
+	else
+		gemmixer(G, user)
+
+/obj/item/precious/gemstone/AltClick(mob/user)
+	if(src.gemsize == GEM_ENORMOUS)
+		user.put_in_hands(src)
+		return
+	if(src.stacksize == 1) // just pick it up...
+		user.put_in_hands(src)
+		return
+	if(src.gemtype == GEM_MIXED)
+		to_chat(user, span("notice", "Try disassembling the whole stack instead, if you want to separate it."))
+		return
+	else 
+		to_chat(user, span("notice", "You take a [src.gemsize] [src.gemtype] from the pile."))
+		src.stacksize -= 1
+		src.update_gem()
+		var/obj/O = new type(loc, 1)
+		user.put_in_hands(O)
+		src.update_gem()
+
+/obj/item/precious/gemstone/attack_self(mob/user)
+	. = ..()
+	if(src.gemsize == GEM_ENORMOUS)
+		user.visible_message(span("notice", "[user] admires the sparkling facets of [src]."),
+		span("notice", "You gaze into the sparkling facets of [src.name]. Hypnotically beautiful..."))
+		return 
+	switch (stacksize)
+		if(1)
+			user.visible_message(span("notice", "[user] flips [src] in one hand."),
+			span("notice", "You give [src] a little flip. Shiny!"))
+			return 
+		if(1 to 5)
+			user.visible_message(span("notice", "[user] sifts [src] through their fingers, admiring the gleaming gemstones."),
+			span("notice", "You sift [src] through your fingers. Glittery..."),
+			span("notice", "You hear soft clinking."))
+			return 
+		if(6 to 10)
+			user.visible_message(span("notice", "[user] counts [src], looking pleased with their bounty."),
+			span("notice", "You count [src]. Yep, still [stacksize] gems in the pile!"),
+			span("notice", "You hear soft clinking."))
+			return 
+		if(11 to 30)
+			user.visible_message(span("notice", "[user] begins counting [src], looking extremely pleased."),
+			span("notice", "You start counting [src]. 1.. 2.. 3.."),
+			span("notice", "You hear soft clinking."))
+			if(do_after(5))
+				to_chat(user, span("notice", "You finish counting. There's still [stacksize] gems in the pile. Hurray!"))
+			else
+				to_chat(user, span("notice", "You lost your count of [src]!"))
```

Comment:
Unless I'm wrong this would tell you "You lost your count of small emerald", no? Which is a pretty ugly sentence.

Reply:
It should display as 'the pile of small emeralds'